artiq/applets/big_number.py

```diff
@@ -26,47 +27,78 @@ def __init__(self, args, ctl):
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
         self.edit_widget.editCancelled.connect(self.cancel_edit)
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
         font = QtGui.QFont()
         font.setPointSize(60)
+        self.display_widget.setFont(font)
         self.edit_widget.setFont(font)
 
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
+
+    def get_unscaled_format(self):
+        scale = self.metadata.get("scale", None)
+        unit = self.metadata.get("unit", "")
+        unit_scale = scale_from_metadata({"unit": unit})
+        if scale is not None:
+            value = self.value/scale
+        else:
+            value = self.value/unit_scale
+        return f"{value} {unit}".strip()
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value == 0:
+            self.edit_widget.setText("---")
+        else:
+            self.edit_widget.setText(self.get_unscaled_format())
         self.edit_widget.selectAll()
         self.edit_widget.setFocus()
         self.setCurrentWidget(self.edit_widget)
 
     def confirm_edit(self):
-        value = float(self.edit_widget.text())
-        self.ctl.set_dataset(self.dataset_name, value)
-        self.setCurrentWidget(self.lcd_widget)
+        text = self.edit_widget.text().split()
+        try:
+            value = float(text[0])
+        except ValueError:
+            self.display_widget.setText("---")
+            self.setCurrentWidget(self.display_widget)
+            return
+        unit = text[1] if len(text) > 1 else ""
+        scale = scale_from_metadata({"unit": unit})
+        value = value*scale
+        precision = self.metadata.get("precision", None)
+        self.ctl.set_dataset(self.dataset_name, value, unit=unit,
+                             precision=precision)
+        self.setCurrentWidget(self.display_widget)
 
     def cancel_edit(self):
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
 
     def data_changed(self, value, metadata, persist, mods):
+        self.metadata = metadata.get(self.dataset_name, {})
         try:
-            n = float(value[self.dataset_name])
+            self.value = float(value[self.dataset_name])
+            n = short_format(self.value, self.metadata)
```

Comment:
related to my previous comment, why not store `n` as a member variable and pass it straight to the edit widget? just need to make sure the submission converts everything properly

Reply:
edit: need slightly different display, so cannot do this.